### PR TITLE
SB-122 Create built-in privileges

### PR DESCRIPTION
**Pull request by AlexOreshkevich**
Hi Martin,

ive added security to several APIs to test how roles working, but it seems like you manage differently authorisation process for `ConfigurationManagementRestletTest` and for `ArtifactRestletTest`. Last test used secured API endpoints, but it's working. First test used also secured endpoints, but it's not working.

Could you take a look please?

```
Failed tests: 
  ConfigurationManagementRestletTest.testAddGetStorage:101 Failed to create storage! expected:<200> but was:<401>
  ConfigurationManagementRestletTest.testCreateAndDeleteStorage:146 Failed to create storage! expected:<200> but was:<401>
  ConfigurationManagementRestletTest.testSetAndGetBaseUrl:58 Failed to set baseUrl! expected:<200> but was:<401>
  ConfigurationManagementRestletTest.testSetAndGetGlobalProxyConfiguration:78 Failed to set proxy configuration! expected:<200> but was:<401>
  ConfigurationManagementRestletTest.testSetAndGetPort:46 Failed to set port! expected:<200> but was:<401>
Tests in error: 
  ArtifactRestletTest.testResolveViaProxy:130->createProxiedRepository:238 » NullPointer
  ConfigurationManagementRestletTest.testGetAndSetConfiguration:207 NullPointer

Tests run: 31, Failures: 5, Errors: 2, Skipped: 0
```

PS Also review my implementation of roles and privileges.

### CODE REVIEW

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/rest/ConfigurationManagementRestlet.java [hunk 1]
```diff
@@ -30,6 +31,7 @@
 @Component
 @Path("/configuration/strongbox")
 @Api(value = "/configuration/strongbox")
+@PreAuthorize("hasRole('ARTIFACTS_DELETE')")
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/rest/ConfigurationManagementRestlet.java]: @AlexOreshkevich ,
I am not quite sure why you need the permissions to delete artifacts in this restlet. It's actually meant for handling the configuration of the server.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/rest/LoggingManagementRestlet.java [hunk 2]
```diff
@@ -1,21 +1,23 @@
...
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Component;
+
 /**
  * @author carlspring
  */
 @Component
 @Path("/logging")
 @Api(value = "/logging")
+@PreAuthorize("permitAll")
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/rest/LoggingManagementRestlet.java]: @AlexOreshkevich ,
I  think this needs to be furter fine-grained. Not everyone will be allowed to configure or view logs.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/rest/MetadataManagementRestlet.java [hunk 3]
```diff
@@ -22,12 +16,21 @@
...
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Component;
+
 /**
  * @author Martin Todorov
  */
 @Component
 @Path("/metadata")
 @Api(value = "/metadata")
+@PreAuthorize("permitAll")
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/rest/MetadataManagementRestlet.java]: @AlexOreshkevich ,
I think this needs to be furter fine-grained. Not everyone will be allowed to rebuild the metadata.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/rest/SearchRestlet.java [hunk 4]
```diff
@@ -21,9 +15,18 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Component;
+
 @Component
 @Path("/search")
 @Api(value = "/search")
+@PreAuthorize("permitAll")
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/rest/SearchRestlet.java]: @AlexOreshkevich ,
I think this needs to be furter fine-grained. Not everyone will be allowed to search.

File: strongbox-web-core/src/main/java/org/carlspring/strongbox/rest/TrashRestlet.java [hunk 5]
```diff
@@ -1,25 +1,28 @@
...
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.stereotype.Component;
+
 /**
  * @author Martin Todorov
  */
 @Component
 @Path("/trash")
 @Api(value = "/trash")
+@PreAuthorize("permitAll")
```
Comment (carlspring) [inline strongbox-web-core/src/main/java/org/carlspring/strongbox/rest/TrashRestlet.java]: @AlexOreshkevich ,
I think this needs to be furter fine-grained. Not everyone will be allowed to clear the trash. There will have to be a privilege for this as well.

File: strongbox-client/src/main/java/org/carlspring/strongbox/client/ArtifactClient.java [hunk 6]
```diff
@@ -278,19 +295,26 @@ public void delete(String storageId,
                        boolean force)
             throws ArtifactOperationException
     {
-        @SuppressWarnings("ConstantConditions")
-        String url = getContextBaseUrl() + "/storages/" + storageId + "/" + repositoryId + "/" + path +
-                     (force ? "?force=" + force : "");
+        try
```
Comment (carlspring) [inline strongbox-client/src/main/java/org/carlspring/strongbox/client/ArtifactClient.java]: @AlexOreshkevich : I don't think I quite agree with the `try`/`catch` here... Could you please remove it, unless there is a good reason I'm missing?

Comment (AlexOreshkevich) [inline strongbox-client/src/main/java/org/carlspring/strongbox/client/ArtifactClient.java]: try-catch removed, Ive added it for testing purpose

File: strongbox-user-management/strongbox-user-management-api/src/main/java/org/carlspring/strongbox/users/domain/Privileges.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    UI_MANAGER(Roles.uiAll());
+
+    private Set<Roles> roles;
+
+    /**
+     * Define custom privilege based on one or multiple roles.
+     *
+     * @param roles
+     */
+    private Privileges(Roles... roles)
```
Comment (carlspring) [inline strongbox-user-management/strongbox-user-management-api/src/main/java/org/carlspring/strongbox/users/domain/Privileges.java]: @AlexOreshkevich,

I'm not quite sure what this about...
`Role`s can contain privileges, or other roles.

Here is an example:
* Privileges:
  * Logging: Read logs
  * Logging: Configure
* Roles
  * Logging: Full
    * Logging: Read logs
    * Logging: Configure
  * Administrator
    * Logging: Full
    * Role 1
    * Role 2
    * Privilege 1
    * Privilege 2

**Pull request merged**